src/misc/unicode.cpp

```diff
@@ -0,0 +1,1129 @@
+
+    LOG_ERR("UTF8: Error, invalid grapheme defined in file %s, line %d",
+            file_name.c_str(), static_cast<int>(line_num));
+    return false;    
+}
+
+static const std::regex regex_map_ascii(
+    // hex_4_digits NNN|CPS|HSH|character
+    "^0x([[:xdigit:]]{4})[[:space:]]+.*(NNN|SPC|HSH|[^[:space:]])$",
+    std::regex_constants::optimize);
+
+static const std::regex regex_map_code(
+    // hex_2_digits hex_4_digits
+    "^0x([[:xdigit:]]{2})[[:space:]]+0x([[:xdigit:]]{4})$",
+    std::regex_constants::optimize);
+
+static const std::regex regex_map_code_mark1(
+    // hex_2_digits hex_4_digits
+    "^0x([[:xdigit:]]{2})[[:space:]]+0x([[:xdigit:]]{4})[[:space:]]+0x([[:xdigit:]]{4})$",
+    std::regex_constants::optimize);
+
+static const std::regex regex_map_code_mark2(
+    // hex_2_digits hex_4_digits
+    "^0x([[:xdigit:]]{2})[[:space:]]+0x([[:xdigit:]]{4})[[:space:]]+0x([[:xdigit:]]{4})[[:space:]]+0x([[:xdigit:]]{4})$",
+    std::regex_constants::optimize);
+
+static const std::regex regex_map_undefined(
+    // hex_2_digits hex_4_digits
+    "^0x[[:xdigit:]]{2}$",
+    std::regex_constants::optimize);
+
+static const std::regex regex_map_alias(
+    // ALIAS hex_4_digits hex_4_digits [BIDIRECTIONAL]
+    "^ALIAS[[:space:]]+0x([[:xdigit:]]{4})[[:space:]]+0x([[:xdigit:]]{4})([[:space:]]+BIDIRECTIONAL)?$",
+    std::regex_constants::optimize);
+
+static const std::regex regex_cp_start(
+    // CODEPAGE dec
+    "^CODEPAGE[[:space:]]+([[:digit:]]{1,5})$",
+    std::regex_constants::optimize);
+
+static const std::regex regex_cp_duplicates(
+    // CODEPAGE dec DUPLICATES dec
+    "^CODEPAGE[[:space:]]+([[:digit:]]{1,5})[[:space:]]+DUPLICATES[[:space:]]+([[:digit:]]{1,5})$",
+    std::regex_constants::optimize);
+
+static const std::regex regex_extend_file(
+    // EXTENDS FILE dir_name file_NAME
+    "^EXTENDS[[:space:]]+FILE[[:space:]]+([^[:space:]]+)[[:space:]]+([^[:space:]]+)$",
+    std::regex_constants::optimize);
+
+static const std::regex regex_extend_cp(
+    // EXTENDS CODEPAGE dec
+    "^EXTENDS[[:space:]]+CODEPAGE[[:space:]]+([[:digit:]]+)$",
+    std::regex_constants::optimize);
+
+
+static bool import_mapping_code_page(const std_fs::path &path_root,
+                                     const std::string &file_name,
+                                     CodePageMappingReverse &mapping)
+{
+    // Import code page character -> UTF-8 mapping from external file
+
+    // Open the file
+    auto in_file = open_mapping_file(path_root, file_name);
+    if (!in_file) {
+        LOG_ERR("UTF8: Error opening mapping file %s", file_name.c_str());
+        return false;
+    }
+
+    // Read and parse
+    std::string line_str = " ";
+    size_t      line_num = 0;
+    std::smatch match;
+
+    CodePageMappingReverse new_mapping;
+
+    while (get_mapping_line(in_file, line_str, line_num)) {
+
+        if (line_str.length() == 1 && line_str[0] == 0x1a)
+            break; // end of file marking, for some fiels from unicode.org
```

Comment:
typo: fiels
How about extracting `0x1a` into a constant?

Reply:
Done, for all such constants.